File: apps/worker/src/aurum_worker/models/trading.py
"""Read-side broker and proposal wire contracts."""

from __future__ import annotations

from enum import StrEnum
from typing import Annotated, Literal
from uuid import UUID

from pydantic import (
    AwareDatetime,
    Field,
    StringConstraints,
    TypeAdapter,
    model_validator,
)

from aurum_worker.models.base import (
    CurrencyCode,
    FiniteFloat,
    Identifier,
    MaximumPermittedVolume,
    NonNegativeInt,
    PositiveFloat,
    PositiveInt,
    WireModel,
)
from aurum_worker.models.eligibility import EligibilityPolicyResult

PermittedVolume = Annotated[float, Field(gt=0, le=0.01, allow_inf_nan=False)]
TokenHash = Annotated[
    str,
    StringConstraints(strict=True, min_length=32, max_length=256),
]


class BrokerSymbolSpecification(WireModel):
    canonical_symbol: Literal["XAUUSD"]
    broker_symbol: Identifier
    specification_version: Identifier
    account_currency: CurrencyCode
    contract_size: PositiveFloat
    digits: NonNegativeInt
    point_size: PositiveFloat
    tick_size: PositiveFloat
    tick_value: PositiveFloat | None
    minimum_volume: PositiveFloat
    maximum_volume: PositiveFloat
    volume_step: PositiveFloat
    stop_level: NonNegativeInt
    calculation_mode: Identifier
    fetched_at: AwareDatetime

    @model_validator(mode="after")
    def validate_volume_range(self) -> BrokerSymbolSpecification:
        if self.minimum_volume > self.maximum_volume:
            raise ValueError(
                "Broker minimum volume cannot exceed broker maximum volume."
            )
        return self


class PositionSizingCalculationSource(StrEnum):
    MT5_ORDER_CALC_PROFIT = "mt5_order_calc_profit"
    BROKER_TICK_VALUE = "broker_tick_value"
    SIMULATION = "simulation"


class _PositionSizingBase(WireModel):
    entry_price: PositiveFloat
    stop_loss_price: PositiveFloat
    stop_distance_price: PositiveFloat
    stop_distance_points: PositiveFloat
    account_equity: PositiveFloat
    risk_limit_pct: PositiveFloat
    risk_budget_amount: PositiveFloat
    calculated_volume: PositiveFloat
    broker_minimum_volume: PositiveFloat
    broker_volume_step: PositiveFloat
    maximum_permitted_volume: MaximumPermittedVolume
    estimated_loss_at_stop: PositiveFloat
    actual_risk_pct: PositiveFloat
    unused_risk_capacity: FiniteFloat
    calculation_source: PositionSizingCalculationSource

    @model_validator(mode="after")
    def validate_reconciliation(self) -> _PositionSizingBase:
        tolerance = 0.011
        expected_distance = abs(self.entry_price - self.stop_loss_price)
        if abs(expected_distance - self.stop_distance_price) > tolerance:
            raise ValueError("Stop distance must match entry and Stop Loss prices.")

        expected_budget = self.account_equity * (self.risk_limit_pct / 100)
        if abs(expected_budget - self.risk_budget_amount) > tolerance:
            raise ValueError(
                "Risk budget must reconcile with equity and risk percentage."
            )

        expected_risk = (self.estimated_loss_at_stop / self.account_equity) * 100
        if abs(expected_risk - self.actual_risk_pct) > tolerance:
            raise ValueError(
                "Actual risk must reconcile with estimated loss and equity."
            )
        return self


class PositionSizingPass(_PositionSizingBase):
    result: Literal["pass"]
    requested_volume: PermittedVolume
    approved_volume: PermittedVolume | None

    @model_validator(mode="after")
    def validate_approval_ceiling(self) -> PositionSizingPass:
        if (
            self.approved_volume is not None
            and self.approved_volume > self.requested_volume
        ):
            raise ValueError("Approved volume cannot exceed requested volume.")
        return self


class PositionSizingBlock(_PositionSizingBase):
    result: Literal["block"]
    requested_volume: None
    approved_volume: None
    block_reason: Identifier


type PositionSizingResult = Annotated[
    PositionSizingPass | PositionSizingBlock,
    Field(discriminator="result"),
]
POSITION_SIZING_ADAPTER: TypeAdapter[PositionSizingResult] = TypeAdapter(
    PositionSizingResult
)


class TradeProposalStatus(StrEnum):
    CANDIDATE = "candidate"
    VALIDATED = "validated"
    PENDING_APPROVAL = "pending_approval"
    APPROVED = "approved"
    REJECTED = "rejected"
    BLOCKED = "blocked"
    EXPIRED = "expired"
    EXECUTION_PENDING = "execution_pending"
    EXECUTED = "executed"
    FAILED = "failed"


class TradeDirection(StrEnum):
    BUY = "BUY"
    SELL = "SELL"


class TradeProposal(WireModel):
    id: UUID
    proposal_version: PositiveInt
    user_id: UUID
    trading_account_id: UUID
    account_type: Literal["demo"]
    account_currency: CurrencyCode
    broker_server: Identifier
    canonical_symbol: Literal["XAUUSD"]
    broker_symbol: Identifier
    symbol_specification_version: Identifier
    direction: TradeDirection
    strategy_code: Identifier
    strategy_version: Identifier
    model_version: Identifier | None = None
    eligibility_policy_version: Identifier
    risk_policy_version: Identifier
    entry_price: PositiveFloat
    stop_loss_price: PositiveFloat
    take_profit_price: PositiveFloat
    calculated_volume: PositiveFloat
    requested_volume: PermittedVolume | None
    approved_volume: PermittedVolume | None
    maximum_permitted_volume: MaximumPermittedVolume
    risk_amount: PositiveFloat
    risk_pct: PositiveFloat
    risk_reward: PositiveFloat
    market_snapshot_id: UUID
    feature_snapshot_id: UUID
    decision_trace_id: UUID
    eligibility: EligibilityPolicyResult
    status: TradeProposalStatus
    created_at: AwareDatetime
    expires_at: AwareDatetime
    processed_at: AwareDatetime | None = None

    @model_validator(mode="after")
    def validate_trade_invariants(self) -> TradeProposal:
        if self.direction is TradeDirection.BUY:
            prices_are_ordered = (
                self.stop_loss_price < self.entry_price < self.take_profit_price
            )
        else:
            prices_are_ordered = (
                self.take_profit_price < self.entry_price < self.stop_loss_price
            )
        if not prices_are_ordered:
            raise ValueError(
                "Entry, Stop Loss, and Take Profit are invalid for direction."
            )

        if self.created_at >= self.expires_at:
            raise ValueError("Proposal expiry must be after creation.")
        if self.status is TradeProposalStatus.BLOCKED and (
            self.requested_volume is not None or self.approved_volume is not None
        ):
            raise ValueError("Blocked proposals cannot request or approve volume.")
        if self.approved_volume is not None and self.requested_volume is None:
            raise ValueError("Approved volume requires a requested volume.")
        if (
            self.approved_volume is not None
            and self.requested_volume is not None
            and self.approved_volume > self.requested_volume
        ):
            raise ValueError("Approved volume cannot exceed requested volume.")
        return self


class PersistedTradeProposal(WireModel):
    """SQL proposal summary without unpersisted eligibility-check evidence."""

    id: UUID
    owner_id: UUID
    proposal_version: PositiveInt
    trading_account_id: UUID
    broker_symbol_id: UUID
    risk_policy_version_id: UUID
    account_type: Literal["demo"]
    account_currency: CurrencyCode
    broker_server: Identifier
    canonical_symbol: Literal["XAUUSD"]
    broker_symbol: Identifier
    symbol_specification_version: Identifier
    direction: TradeDirection
    strategy_code: Identifier
    strategy_version: Identifier
    model_version: Identifier | None
    eligibility_policy_id: Identifier
    eligibility_policy_version: Identifier
    eligibility_outcome: Literal["auto", "ask", "block"]
    eligibility_evaluated_at: AwareDatetime
    risk_policy_version: Identifier
    entry_price: PositiveFloat
    stop_loss_price: PositiveFloat
    take_profit_price: PositiveFloat
    calculated_volume: PositiveFloat
    requested_volume: PermittedVolume | None
    approved_volume: PermittedVolume | None
    maximum_permitted_volume: MaximumPermittedVolume
    risk_amount: PositiveFloat
    risk_pct: PositiveFloat
    risk_reward: PositiveFloat
    market_snapshot_id: UUID
    feature_snapshot_id: UUID
    decision_trace_id: UUID
    status: TradeProposalStatus
    created_at: AwareDatetime
    expires_at: AwareDatetime
    processed_at: AwareDatetime | None
    updated_at: AwareDatetime

    @model_validator(mode="after")
    def validate_persisted_summary(self) -> PersistedTradeProposal:
        if self.direction is TradeDirection.BUY:
            prices_are_ordered = (
                self.stop_loss_price < self.entry_price < self.take_profit_price
            )
        else:
            prices_are_ordered = (
                self.take_profit_price < self.entry_price < self.stop_loss_price
            )
        if not prices_are_ordered:
            raise ValueError(
                "Entry, Stop Loss, and Take Profit are invalid for direction."
            )
        if self.created_at >= self.expires_at:
            raise ValueError("Proposal expiry must be after creation.")
        if self.status is TradeProposalStatus.BLOCKED and (
            self.requested_volume is not None or self.approved_volume is not None
        ):
            raise ValueError("Blocked proposals cannot request or approve volume.")
        if self.approved_volume is not None and self.requested_volume is None:
            raise ValueError("Approved volume requires a requested volume.")
        if (
            self.approved_volume is not None
            and self.requested_volume is not None
            and self.approved_volume > self.requested_volume
        ):
            raise ValueError("Approved volume cannot exceed requested volume.")
        return self


class MobileApprovalSessionStatus(StrEnum):
    CREATED = "created"
    VALIDATED = "validated"
    USED = "used"
    EXPIRED = "expired"
    REVOKED = "revoked"


class MobileApprovalSession(WireModel):
    id: UUID
    proposal_id: UUID
    proposal_version: PositiveInt
    allowed_user_id: Identifier
    token_hash: TokenHash
    nonce: Identifier
    status: MobileApprovalSessionStatus
    created_at: AwareDatetime
    expires_at: AwareDatetime
    used_at: AwareDatetime | None = None

    @model_validator(mode="after")
    def validate_session_lifecycle(self) -> MobileApprovalSession:
        if self.created_at >= self.expires_at:
            raise ValueError("Approval session expiry must be after creation.")
        if self.status is MobileApprovalSessionStatus.USED and self.used_at is None:
            raise ValueError("Used approval sessions require a usedAt timestamp.")
        return self

File: packages/contracts/src/trading.ts
import { z } from "zod";

import { EligibilityPolicyResultSchema } from "./eligibility";
import {
  CurrencyCodeSchema,
  IdentifierSchema,
  IsoDateTimeSchema,
  PositiveNumberSchema,
  UuidSchema,
  isBefore,
} from "./primitives";

export const BrokerSymbolSpecificationSchema = z
  .object({
    canonicalSymbol: z.literal("XAUUSD"),
    brokerSymbol: IdentifierSchema,
    specificationVersion: IdentifierSchema,
    accountCurrency: CurrencyCodeSchema,
    contractSize: PositiveNumberSchema,
    digits: z.number().int().nonnegative(),
    pointSize: PositiveNumberSchema,
    tickSize: PositiveNumberSchema,
    tickValue: PositiveNumberSchema.nullable(),
    minimumVolume: PositiveNumberSchema,
    maximumVolume: PositiveNumberSchema,
    volumeStep: PositiveNumberSchema,
    stopLevel: z.number().int().nonnegative(),
    calculationMode: IdentifierSchema,
    fetchedAt: IsoDateTimeSchema,
  })
  .strict()
  .refine(
    (specification) =>
      specification.minimumVolume <= specification.maximumVolume,
    {
      path: ["minimumVolume"],
      message: "Broker minimum volume cannot exceed broker maximum volume.",
    },
  );

export type BrokerSymbolSpecification = z.infer<
  typeof BrokerSymbolSpecificationSchema
>;

const sizingCommonShape = {
  entryPrice: PositiveNumberSchema,
  stopLossPrice: PositiveNumberSchema,
  stopDistancePrice: PositiveNumberSchema,
  stopDistancePoints: PositiveNumberSchema,
  accountEquity: PositiveNumberSchema,
  riskLimitPct: PositiveNumberSchema,
  riskBudgetAmount: PositiveNumberSchema,
  calculatedVolume: PositiveNumberSchema,
  brokerMinimumVolume: PositiveNumberSchema,
  brokerVolumeStep: PositiveNumberSchema,
  maximumPermittedVolume: z.literal(0.01),
  estimatedLossAtStop: PositiveNumberSchema,
  actualRiskPct: PositiveNumberSchema,
  unusedRiskCapacity: z.number().finite(),
  calculationSource: z.enum([
    "mt5_order_calc_profit",
    "broker_tick_value",
    "simulation",
  ]),
};

const positionSizingUnion = z.discriminatedUnion("result", [
  z
    .object({
      ...sizingCommonShape,
      result: z.literal("pass"),
      requestedVolume: PositiveNumberSchema.max(0.01),
      approvedVolume: PositiveNumberSchema.max(0.01).nullable(),
    })
    .strict(),
  z
    .object({
      ...sizingCommonShape,
      result: z.literal("block"),
      requestedVolume: z.null(),
      approvedVolume: z.null(),
      blockReason: IdentifierSchema,
    })
    .strict(),
]);

export const PositionSizingResultSchema = positionSizingUnion.superRefine(
  (sizing, context) => {
    const tolerance = 0.011;
    const expectedDistance = Math.abs(sizing.entryPrice - sizing.stopLossPrice);
    if (Math.abs(expectedDistance - sizing.stopDistancePrice) > tolerance) {
      context.addIssue({
        code: "custom",
        path: ["stopDistancePrice"],
        message: "Stop distance must match entry and Stop Loss prices.",
      });
    }

    const expectedBudget = sizing.accountEquity * (sizing.riskLimitPct / 100);
    if (Math.abs(expectedBudget - sizing.riskBudgetAmount) > tolerance) {
      context.addIssue({
        code: "custom",
        path: ["riskBudgetAmount"],
        message: "Risk budget must reconcile with equity and risk percentage.",
      });
    }

    const expectedRisk =
      (sizing.estimatedLossAtStop / sizing.accountEquity) * 100;
    if (Math.abs(expectedRisk - sizing.actualRiskPct) > tolerance) {
      context.addIssue({
        code: "custom",
        path: ["actualRiskPct"],
        message: "Actual risk must reconcile with estimated loss and equity.",
      });
    }

    if (
      sizing.result === "pass" &&
      sizing.approvedVolume !== null &&
      sizing.approvedVolume > sizing.requestedVolume
    ) {
      context.addIssue({
        code: "custom",
        path: ["approvedVolume"],
        message: "Approved volume cannot exceed requested volume.",
      });
    }
  },
);

export type PositionSizingResult = z.infer<typeof PositionSizingResultSchema>;

export const TRADE_PROPOSAL_STATUSES = [
  "candidate",
  "validated",
  "pending_approval",
  "approved",
  "rejected",
  "blocked",
  "expired",
  "execution_pending",
  "executed",
  "failed",
] as const;

export const TradeProposalStatusSchema = z.enum(TRADE_PROPOSAL_STATUSES);

export const TRADE_DIRECTIONS = ["BUY", "SELL"] as const;
export const TradeDirectionSchema = z.enum(TRADE_DIRECTIONS);

export const TradeProposalSchema = z
  .object({
    id: UuidSchema,
    proposalVersion: z.number().int().positive(),
    userId: UuidSchema,
    tradingAccountId: UuidSchema,
    accountType: z.literal("demo"),
    accountCurrency: CurrencyCodeSchema,
    brokerServer: IdentifierSchema,
    canonicalSymbol: z.literal("XAUUSD"),
    brokerSymbol: IdentifierSchema,
    symbolSpecificationVersion: IdentifierSchema,
    direction: TradeDirectionSchema,
    strategyCode: IdentifierSchema,
    strategyVersion: IdentifierSchema,
    modelVersion: IdentifierSchema.optional(),
    eligibilityPolicyVersion: IdentifierSchema,
    riskPolicyVersion: IdentifierSchema,
    entryPrice: PositiveNumberSchema,
    stopLossPrice: PositiveNumberSchema,
    takeProfitPrice: PositiveNumberSchema,
    calculatedVolume: PositiveNumberSchema,
    requestedVolume: PositiveNumberSchema.max(0.01).nullable(),
    approvedVolume: PositiveNumberSchema.max(0.01).nullable(),
    maximumPermittedVolume: z.literal(0.01),
    riskAmount: PositiveNumberSchema,
    riskPct: PositiveNumberSchema,
    riskReward: PositiveNumberSchema,
    marketSnapshotId: UuidSchema,
    featureSnapshotId: UuidSchema,
    decisionTraceId: UuidSchema,
    eligibility: EligibilityPolicyResultSchema,
    status: TradeProposalStatusSchema,
    createdAt: IsoDateTimeSchema,
    expiresAt: IsoDateTimeSchema,
    processedAt: IsoDateTimeSchema.optional(),
  })
  .strict()
  .superRefine((proposal, context) => {
    const orderedPrices =
      proposal.direction === "BUY"
        ? proposal.stopLossPrice < proposal.entryPrice &&
          proposal.entryPrice < proposal.takeProfitPrice
        : proposal.takeProfitPrice < proposal.entryPrice &&
          proposal.entryPrice < proposal.stopLossPrice;
    if (!orderedPrices) {
      context.addIssue({
        code: "custom",
        path: ["stopLossPrice"],
        message: "Entry, Stop Loss, and Take Profit are invalid for direction.",
      });
    }

    if (!isBefore(proposal.createdAt, proposal.expiresAt)) {
      context.addIssue({
        code: "custom",
        path: ["expiresAt"],
        message: "Proposal expiry must be after creation.",
      });
    }

    if (
      proposal.status === "blocked" &&
      (proposal.requestedVolume !== null || proposal.approvedVolume !== null)
    ) {
      context.addIssue({
        code: "custom",
        path: ["requestedVolume"],
        message: "Blocked proposals cannot request or approve volume.",
      });
    }

    if (proposal.approvedVolume !== null && proposal.requestedVolume === null) {
      context.addIssue({
        code: "custom",
        path: ["approvedVolume"],
        message: "Approved volume requires a requested volume.",
      });
    }
    if (
      proposal.approvedVolume !== null &&
      proposal.requestedVolume !== null &&
      proposal.approvedVolume > proposal.requestedVolume
    ) {
      context.addIssue({
        code: "custom",
        path: ["approvedVolume"],
        message: "Approved volume cannot exceed requested volume.",
      });
    }
  });

export type TradeProposal = z.infer<typeof TradeProposalSchema>;

/**
 * SQL-aligned proposal summary. It intentionally does not pretend to be a full
 * TradeProposal: the database row stores eligibility summary fields, not the
 * EligibilityPolicyResult.checks evidence required by TradeProposalSchema.
 */
export const PersistedTradeProposalSchema = z
  .object({
    id: UuidSchema,
    ownerId: UuidSchema,
    proposalVersion: z.number().int().positive(),
    tradingAccountId: UuidSchema,
    brokerSymbolId: UuidSchema,
    riskPolicyVersionId: UuidSchema,
    accountType: z.literal("demo"),
    accountCurrency: CurrencyCodeSchema,
    brokerServer: IdentifierSchema,
    canonicalSymbol: z.literal("XAUUSD"),
    brokerSymbol: IdentifierSchema,
    symbolSpecificationVersion: IdentifierSchema,
    direction: TradeDirectionSchema,
    strategyCode: IdentifierSchema,
    strategyVersion: IdentifierSchema,
    modelVersion: IdentifierSchema.nullable(),
    eligibilityPolicyId: IdentifierSchema,
    eligibilityPolicyVersion: IdentifierSchema,
    eligibilityOutcome: z.enum(["auto", "ask", "block"]),
    eligibilityEvaluatedAt: IsoDateTimeSchema,
    riskPolicyVersion: IdentifierSchema,
    entryPrice: PositiveNumberSchema,
    stopLossPrice: PositiveNumberSchema,
    takeProfitPrice: PositiveNumberSchema,
    calculatedVolume: PositiveNumberSchema,
    requestedVolume: PositiveNumberSchema.max(0.01).nullable(),
    approvedVolume: PositiveNumberSchema.max(0.01).nullable(),
    maximumPermittedVolume: z.literal(0.01),
    riskAmount: PositiveNumberSchema,
    riskPct: PositiveNumberSchema,
    riskReward: PositiveNumberSchema,
    marketSnapshotId: UuidSchema,
    featureSnapshotId: UuidSchema,
    decisionTraceId: UuidSchema,
    status: TradeProposalStatusSchema,
    createdAt: IsoDateTimeSchema,
    expiresAt: IsoDateTimeSchema,
    processedAt: IsoDateTimeSchema.nullable(),
    updatedAt: IsoDateTimeSchema,
  })
  .strict()
  .superRefine((proposal, context) => {
    const orderedPrices =
      proposal.direction === "BUY"
        ? proposal.stopLossPrice < proposal.entryPrice &&
          proposal.entryPrice < proposal.takeProfitPrice
        : proposal.takeProfitPrice < proposal.entryPrice &&
          proposal.entryPrice < proposal.stopLossPrice;
    if (!orderedPrices) {
      context.addIssue({
        code: "custom",
        path: ["stopLossPrice"],
        message: "Entry, Stop Loss, and Take Profit are invalid for direction.",
      });
    }
    if (!isBefore(proposal.createdAt, proposal.expiresAt)) {
      context.addIssue({
        code: "custom",
        path: ["expiresAt"],
        message: "Proposal expiry must be after creation.",
      });
    }
    if (
      proposal.status === "blocked" &&
      (proposal.requestedVolume !== null || proposal.approvedVolume !== null)
    ) {
      context.addIssue({
        code: "custom",
        path: ["requestedVolume"],
        message: "Blocked proposals cannot request or approve volume.",
      });
    }
    if (proposal.approvedVolume !== null && proposal.requestedVolume === null) {
      context.addIssue({
        code: "custom",
        path: ["approvedVolume"],
        message: "Approved volume requires a requested volume.",
      });
    }
    if (
      proposal.approvedVolume !== null &&
      proposal.requestedVolume !== null &&
      proposal.approvedVolume > proposal.requestedVolume
    ) {
      context.addIssue({
        code: "custom",
        path: ["approvedVolume"],
        message: "Approved volume cannot exceed requested volume.",
      });
    }
  });

export type PersistedTradeProposal = z.infer<
  typeof PersistedTradeProposalSchema
>;

export const MobileApprovalSessionStatusSchema = z.enum([
  "created",
  "validated",
  "used",
  "expired",
  "revoked",
]);

export const MobileApprovalSessionSchema = z
  .object({
    id: UuidSchema,
    proposalId: UuidSchema,
    proposalVersion: z.number().int().positive(),
    allowedUserId: IdentifierSchema,
    tokenHash: z.string().min(32).max(256),
    nonce: IdentifierSchema,
    status: MobileApprovalSessionStatusSchema,
    createdAt: IsoDateTimeSchema,
    expiresAt: IsoDateTimeSchema,
    usedAt: IsoDateTimeSchema.optional(),
  })
  .strict()
  .superRefine((session, context) => {
    if (!isBefore(session.createdAt, session.expiresAt)) {
      context.addIssue({
        code: "custom",
        path: ["expiresAt"],
        message: "Approval session expiry must be after creation.",
      });
    }
    if (session.status === "used" && !session.usedAt) {
      context.addIssue({
        code: "custom",
        path: ["usedAt"],
        message: "Used approval sessions require a usedAt timestamp.",
      });
    }
  });

export type MobileApprovalSession = z.infer<typeof MobileApprovalSessionSchema>;
